### ISSUE TITLE
BCC^2 and jacobians - WIP

### ISSUE THREAD

**Issue by proteneer** [proteneer/timemachine#167]
I'm herein describing the spec for our BCC routine, where we BCC the atomic charges. This will also help us implement generalized jacobians from system parameters Q to forcefield parameters P. Currently the Jacobian for dQ/dP is the identity.

BCCs will be the first instance where this will longer be the identity matrix.

Given three atoms set with OEAM1BCCELF10Charges charges:

```
Suppose we have the atomic charges x,y,z on three atoms with connectivity:

1 2 3
x-y-z

We can empirically bond charge correct them using two matching bond patterns p_i matching

(a) x:1-y:2
(b) y:2-z:3

The final charges q on each atom will be

1: (x-a)
2: (y+a-b)
3: (z+b)

When we compute the chain rule of dL/dp_i = dL/dq_i dot dq_i/dp_i

Shapes:

dL/dq_i = [Q]
jacobian: dq_i/dp_i = [Q, P]

The jacobian matrix in the above example is:

       P
     a  b
  x -1  0
Q y +1 -1
  z  0 +1
```

Merging - if we have two disjoint systems (eg protein system/ff and small molecule system/ff) that we merge with system parameter sets {Q1 Q2} and ff parameter sets {P1 P2} with corresponding jacobians {J1 J2}

```
Q3 = np.concat {Q1 Q2}
P3 = np.concat {P1 P2}
J3 = block diagonal with zeros on the off diagonals
```

Explicit Jacobians vs vjps:

We don't want to necessarily have to compute the full QxP matrix, since both Q and P can be proportional to the number of atoms in the system. So we don't necessarily need the full jacobian.

After some thought, it looks like the vjps can be simply concatenated:

```
dL/dP3 = np.concat([dL/dQ1 dot dQ1/dP1, dL/dQ2 dot dQ2/dP2])
```

**Comment by chuantian** [comment]
If we use openFF's bond pattern for BCC assignment, the jacobian matrix will be sparse with most of elements equal to zero. And each column should sum up to zero.

**Comment by proteneer** [comment]
One excellent issue that Jeff from OFF brought up is we need to iterate through the bonds then match. This is because BCCs are asymmetric.

**Comment by chuantian** [comment]
Yeah, every row can have multiple non-zeros which are attributed to multiple bond types.

**Comment by proteneer** [comment]
That's not quite what I meant

**Comment by proteneer** [comment]
Initial version of explicit jacobians:
```
import jax
import jax.numpy as jnp
import numpy as np

from jax.ops import index_add


def bcc(charges, bcc_idxs, bcc_params):
    srcs = bcc_idxs[:, 0]
    dsts = bcc_idxs[:, 1]
        
    # we can't do implicit updates, so we need to use index_ops instead
    zeros_src = index_add(jnp.zeros_like(charges), srcs, -bcc_params)
    zeros_dst = index_add(jnp.zeros_like(charges), dsts, bcc_params)

    new_charges = charges + zeros_src + zeros_dst
    return new_charges
    
partial_charges = np.array([5.1, 2.2, 3.3, 1.5, -5.2, -3.2], dtype=np.float64)

bcc_params = np.array([
    1.5,
    0.3,
   -2.4
])

bcc_idxs = np.array([
    [1, 3],
    [0, 5],
    [3, 2],
])

bcc(partial_charges, bcc_idxs, bcc_params)

bcc_grad_fn = jax.jacfwd(bcc, argnums=(2,))
bcc_grad_fn(partial_charges, bcc_idxs, bcc_params)[0]

>>> DeviceArray([[ 0., -1.,  0.],
             [-1.,  0.,  0.],
             [ 0.,  0.,  1.],
             [ 1.,  0., -1.],
             [ 0.,  0.,  0.],
             [ 0.,  1.,  0.]], dtype=float32)
```

**Comment by proteneer** [comment]
woot, and here's the vjp version, which avoids computing the jacobian entirely:

``` python
import jax
import jax.numpy as jnp
import numpy as np
import functools

from jax.ops import index_add


def bcc(charges, bcc_params, bcc_idxs):
    srcs = bcc_idxs[:, 0]
    dsts = bcc_idxs[:, 1]
        
    # we can't do implicit updates, so we need to use index_ops instead
    zeros_src = index_add(jnp.zeros_like(charges), srcs, -bcc_params)
    zeros_dst = index_add(jnp.zeros_like(charges), dsts, bcc_params)

    new_charges = charges + zeros_src + zeros_dst
    return new_charges
    
partial_charges = np.array([5.1, 2.2, 3.3, 1.5, -5.2, -3.2], dtype=np.float64)

N = partial_charges.shape[0]

bcc_params = np.array([
    1.5,
    0.3,
   -2.4,
    1.2,
])

bcc_idxs = np.array([
    [1, 3],
    [0, 5],
    [3, 2],
    [3, 4],
])

primals = partial_charges, bcc_params, bcc_idxs

bcc(*primals)
bcc_partial = functools.partial(bcc, bcc_idxs=bcc_idxs)

bcc_grad_fn = jax.jacfwd(bcc_partial, argnums=(1,))
jacobian = bcc_grad_fn(partial_charges, bcc_params)[0]

primals, f_vjp = jax.vjp(bcc_partial, partial_charges, bcc_params)
primals_out = np.random.rand(N)
adjoint = f_vjp(primals_out)[1]

explicit_vjp = np.matmul(primals_out, jacobian)
np.testing.assert_almost_equal(adjoint, explicit_vjp)
```

**Comment by proteneer** [comment]
This is *finally* implemented via #211
